From 1d4da5168ece408f7830b5e2fe8459889f454c59 Mon Sep 17 00:00:00 2001
From: Jianqun Xu <jay.xu@rock-chips.com>
Date: Thu, 4 Jan 2018 09:56:34 +0800
Subject: [PATCH] soc/rockchip: rk_fiq_debugger register as platform device

Register rockchip fiq debugger as platform device, also do some
codingstyle for the driver.

Change-Id: I78e6c3365275e3d45e37eee86378c350a18f259d
Signed-off-by: Jianqun Xu <jay.xu@rock-chips.com>
---
 drivers/soc/rockchip/rk_fiq_debugger.c | 48 ++++++++++++++++----------
 1 file changed, 29 insertions(+), 19 deletions(-)

diff --git a/drivers/soc/rockchip/rk_fiq_debugger.c b/drivers/soc/rockchip/rk_fiq_debugger.c
index 165626c449fa..964d26f0dea1 100755
--- a/drivers/soc/rockchip/rk_fiq_debugger.c
+++ b/drivers/soc/rockchip/rk_fiq_debugger.c
@@ -464,7 +464,7 @@ void rk_serial_debug_init(void __iomem *base, phys_addr_t phy_base,
 	if (!pdev) {
 		pr_err("Failed to alloc fiq debugger platform device\n");
 		goto out3;
-	};
+	}
 
 #ifdef CONFIG_FIQ_DEBUGGER_TRUST_ZONE
 	if ((signal_irq > 0) && (serial_hwirq > 0)) {
@@ -538,15 +538,18 @@ void rk_serial_debug_init(void __iomem *base, phys_addr_t phy_base,
 	kfree(t);
 }
 
-static const struct of_device_id ids[] __initconst = {
-	{ .compatible = "rockchip,fiq-debugger" },
-	{}
+#if defined(CONFIG_OF)
+static const struct of_device_id rk_fiqdbg_of_match[] = {
+	{ .compatible = "rockchip,fiq-debugger", },
+	{},
 };
+MODULE_DEVICE_TABLE(of, rk_fiqdbg_of_match);
+#endif
 
-static int __init rk_fiq_debugger_init(void) {
-
+static int __init rk_fiqdbg_probe(struct platform_device *pdev)
+{
 	void __iomem *base;
-	struct device_node *np;
+	struct device_node *np = pdev->dev.of_node;
 	unsigned int id, ok = 0;
 	int irq, signal_irq = -1, wake_irq = -1;
 	unsigned int baudrate = 0, irq_mode = 0;
@@ -557,13 +560,6 @@ static int __init rk_fiq_debugger_init(void) {
 	struct of_phandle_args oirq;
 	struct resource res;
 
-	np = of_find_matching_node(NULL, ids);
-
-	if (!np) {
-		pr_err("fiq-debugger is missing in device tree!\n");
-		return -ENODEV;
-	}
-
 	if (!of_device_is_available(np)) {
 		pr_err("fiq-debugger is disabled in device tree\n");
 		return -ENODEV;
@@ -633,8 +629,22 @@ static int __init rk_fiq_debugger_init(void) {
 				     irq, signal_irq, wake_irq, baudrate);
 	return 0;
 }
-#ifdef CONFIG_FIQ_GLUE
-postcore_initcall_sync(rk_fiq_debugger_init);
-#else
-arch_initcall_sync(rk_fiq_debugger_init);
-#endif
+
+static struct platform_driver rk_fiqdbg_driver = {
+	.driver = {
+		.name   = "rk-fiq-debugger",
+		.of_match_table = of_match_ptr(rk_fiqdbg_of_match),
+	},
+};
+
+static int __init rk_fiqdbg_init(void)
+{
+	return platform_driver_probe(&rk_fiqdbg_driver,
+				     rk_fiqdbg_probe);
+}
+arch_initcall_sync(rk_fiqdbg_init);
+
+MODULE_AUTHOR("Huibin Hong <huibin.hong@rock-chips.com>");
+MODULE_DESCRIPTION("Rockchip FIQ Debugger");
+MODULE_LICENSE("GPL");
+MODULE_ALIAS("platform:rk-fiq-debugger");
-- 
2.35.3

